### PR TITLE
[NO GBP] Resolves Golems not being allowed in the areas they created

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Creates a new signal that fires whenever a new area is created, or expanded, when using the create_area proc.
Also implements this signal in the Hazard Area component to allow new areas that are created to be considered non-hazardous if they meet the criteria for non hazardous areas.

## Why It's Good For The Game

If you build a room we should probably allow you to be in it

## Changelog

:cl:
fix: Golems can now exist in the rooms they themselves create, sorry about that.
/:cl:

### CODE REVIEW

File: code/__DEFINES/dcs/signals/signals_area.dm [hunk 1]
```diff
@@ -13,6 +13,9 @@
 ///from base of area/Exited(): (/area). Sent to "area-sensitive" movables, see __DEFINES/traits.dm for info.
 #define COMSIG_EXIT_AREA "exit_area"
 
+/// Called when a new area is created by a mob using create_area - (area/created_area, area/old_area, mob/creator)
+#define COMSIG_AREA_CREATED "mob_created_area"
```
Comment (Mothblocks) [inline code/__DEFINES/dcs/signals/signals_area.dm:17]: This needs to be formatted like the other global signals, and put in https://github.com/tgstation/tgstation/blob/master/code/__DEFINES/dcs/signals/signals_global.dm

File: code/datums/components/hazard_area.dm [hunk 2]
```diff
@@ -21,19 +23,27 @@
...
 
 /datum/component/hazard_area/UnregisterFromParent()
 	var/mob/parent_mob = parent
 	UnregisterSignal(parent_mob, list(COMSIG_ENTER_AREA, COMSIG_LADDER_TRAVEL, COMSIG_VEHICLE_RIDDEN))
+	UnregisterSignal(SSdcs, COMSIG_AREA_CREATED)
 	parent_mob.lose_area_sensitivity(type)
 
+/datum/component/hazard_area/Destroy(force, silent)
+	. = ..()
+	area_created.Cut() // Lets not have hanging refs to areas thanks
```
Comment (Mothblocks) [inline code/datums/components/hazard_area.dm:44]: Just setting it to null is fine.

File: code/datums/components/hazard_area.dm [hunk 3]
```diff
@@ -71,6 +82,26 @@
...
+	if(created.type in area_whitelist)
+		return // in whitelist, probably expanded an already whitelisted area
+
+	if(created.type in area_blacklist)
+		return // in blacklist, expanding a blacklisted area doesnt magically give you permission to enter
+
+	if(overwritten)
+		if(check_area_hazardous(overwritten.type))
+			return // Overwrote a hazardous area, still hazardous fool
+		area_created -= overwritten // While its not guarenteed to be in the area_created list it's a good idea to ensure we dont have handing refs
```
Comment (Mothblocks) [inline code/datums/components/hazard_area.dm:97]:
```suggestion
		area_created -= overwritten // While its not guaranteed to be in the area_created list it's a good idea to ensure we dont have handing refs
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by ZephyrTFA** [comment]
Need to add survival pods to the area whitelist for golems

**Comment by Learner10** [comment]
Thank you.

**Pull request merged**